cartography/intel/aws/kms.py

```diff
@@ -0,0 +1,333 @@
+        grants.extend(page['Grants'])
+
+    return grants
+
+
+@timeit
+def _load_kms_key_aliases(neo4j_session, aliases, update_tag):
+    """
+    Ingest KMS Aliases into neo4j.
+    """
+    ingest_aliases = """
+    MERGE (a:KMSAlias{id: {AliasArn}})
+    ON CREATE SET a.firstseen = timestamp(), a.targetkeyid = {TargetKeyId}
+    SET a.aliasname = {AliasName}, a.lastupdated = {UpdateTag}
+    WITH a MATCH (kmskey:KMSKey{id: {TargetKeyId}})
+    MERGE (a)-[r:KNOWN_AS]->(kmskey)
+    ON CREATE SET r.firstseen = timestamp()
+    SET r.lastupdated = {UpdateTag}
+    """
+
+    for alias in aliases:
+        neo4j_session.run(
+            ingest_aliases,
+            AliasArn=alias["AliasArn"],
+            AliasName=alias.get("AliasName"),
+            TargetKeyId=alias["TargetKeyId"],
+            UpdateTag=update_tag,
+        )
+
+
+@timeit
+def _load_kms_key_grants(neo4j_session, grants_list, update_tag):
+    """
+    Ingest KMS Key Grants into neo4j.
+    """
+    ingest_grants = """
+    MERGE (g:KMSGrant{id: {GrantId}})
+    ON CREATE SET g.firstseen = timestamp(), g.granteeprincipal = {GranteePrincipal}, g.creationdate = {CreationDate}
+    SET g.name = {GrantName}, g.lastupdated = {UpdateTag}
+    WITH g
+    MATCH (kmskey:KMSKey{id: {TargetKeyId}})
+    MERGE (g)-[r:APPLIED_ON]->(kmskey)
+    ON CREATE SET r.firstseen = timestamp()
+    SET r.lastupdated = {UpdateTag}
+    """
+
+    for grant in grants_list:
+        neo4j_session.run(
+            ingest_grants,
+            GrantId=grant["GrantId"],
+            GrantName=grant.get("Name"),
+            GranteePrincipal=grant.get("GranteePrincipal"),
+            CreationDate=grant["CreationDate"],
+            TargetKeyId=grant["KeyId"],
+            UpdateTag=update_tag,
+        )
+
+
+@timeit
+def _load_kms_key_policies(neo4j_session, policies, update_tag):
+    """
+    Ingest KMS Key policy results into neo4j.
+    """
+    # NOTE we use the coalesce function so appending works when the value is null initially
+    ingest_policies = """
+    UNWIND {policies} AS policy
+    MATCH (k:KMSKey) where k.name = policy.kms_key
+    SET k.anonymous_access = (coalesce(k.anonymous_access, false) OR policy.internet_accessible),
+    k.anonymous_actions = coalesce(k.anonymous_actions, []) + policy.accessible_actions,
+    k.lastupdated = {UpdateTag}
+    """
+
+    neo4j_session.run(
+        ingest_policies,
+        policies=policies,
+        UpdateTag=update_tag,
+    )
+
+
+def _set_default_values(neo4j_session, aws_account_id):
+    set_defaults = """
+    MATCH (:AWSAccount{id: {AWS_ID}})-[:RESOURCE]->(kmskey:KMSKey) where NOT EXISTS(kmskey.anonymous_actions)
+    SET kmskey.anonymous_access = false, kmskey.anonymous_actions = []
+    """
+
+    neo4j_session.run(
+        set_defaults,
+        AWS_ID=aws_account_id,
+    )
+
+
+@timeit
+def load_kms_key_details(neo4j_session, policy_alias_grants_data, region, aws_account_id, update_tag):
```

Comment:
I realize you used the s3 sync as a model here, and while this works, I wonder if using Resource Permission Relationships would be more powerful. See https://github.com/lyft/cartography/blob/master/docs/usage/permissions-mapping.md and (for a longer read) https://eng.lyft.com/iam-whatever-you-say-iam-febce59d1e3b.

The rough idea is that you edit the permission_relationships.yaml file and cartography draws links from the IAM principals to the resource nodes based on the allowed actions:
![image](https://user-images.githubusercontent.com/46503781/107660728-6cda9000-6c3d-11eb-96f7-a7e0433545a5.png)

This RPR change for KMS keys might not get merged here depending on how it impacts performance (which is why we have that yaml config file that lets users decide which permissions they care about syncing) but I wanted to let you know that this feature exists.

Reply:
@achantavy you are right, we used S3 as the reference while populating KMS entities into Cartography. Our idea is to capture all the policies, statements, etc. which would allow the users for more fine grained queries